OpInfo: div - port pending method_tests entry

kshitij12345 opened the pull request:
Reference: #54261 

Depends on: #59154

kshitij12345 committed:
OpInfo: true_divide and minor fix

kshitij12345 committed:
add to works_list

kshitij12345 committed:
add comment

kshitij12345 committed:
OpInfo: div - port pending method_tests entry

kshitij12345 committed:
add link to the issue

kshitij12345 committed:
Merge branch 'master' into develop/opinfo/div

kshitij12345 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -4701,26 +4693,30 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
            dtypes=all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16),
-           sample_inputs_func=partial(sample_inputs_div, rounding_mode=None),
-           skips=(SkipInfo('TestOpInfo', 'test_duplicate_method_tests'),),
+           sample_inputs_func=partial(sample_inputs_binary_pwise, rounding_mode=None, rhs_exclude_zero=True),
```
This generates the same sample as above case. Should we just remove either one of them?

mruberry replied:
I agree that this OpInfo seems redundant with the above and can be eliminated

kshitij12345 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -146,7 +146,7 @@ def formatter(arg):
             elif is_iterable_of_tensors(arg):
                 return "TensorList[" + ", ".join(map(formatter, arg)) + "]"
-            elif is_iter(arg):  # Handle list, tuple or any iterable type
+            elif is_iter(arg) and not isinstance(arg, str):  # Handle list, tuple or any iterable type
```
`str` is also iterable. So we stay in this branch forever without this condition.

mruberry replied:
What about make it more explicit to just handle lists and tuples?

kshitij12345 replied:
That would be fine. But if it is any other iterable holding tensor, we will fall to `repr(arg)` and end up printing the whole tensor.

mruberry replied:
I understand; but we can deal with that later

kshitij12345 replied:
Cool!

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -739,6 +739,7 @@ def sample_inputs_linalg_vector_norm(op_info, device, dtype, requires_grad, **kw
 # specify scalars of floating, integral & complex types as values for "alpha".
 # Keyword argument `rhs_exclude_zero` is used to exclude zero values from rhs tensor argument
 # This is necessary for operations like `true_divide`, where divide by zero throws an exception.
+# Keyword argument `rounding_mode` is specifically used for the same kwarg supported by `torch.div`.
```
Instead of accepting a kwarg just for rounding_mode, what about accepting a dict, extra_args, that's passed as kwargs to the function?

`sample_inputs_binary_pwise(..., *, extra_args={}, **kwargs)`

Maybe `extra_args` should be None instead of an empty dict

kshitij12345 replied:
That makes sense. Thanks!

kshitij12345 replied:
With this change diff will look something like this. Does this look okay?

```patch
diff --git a/torch/testing/_internal/common_methods_invocations.py b/torch/testing/_internal/common_methods_invocations.py
index 474420fb3c..81b8942ec2 100644
--- a/torch/testing/_internal/common_methods_invocations.py
+++ b/torch/testing/_internal/common_methods_invocations.py
@@ -725,8 +725,10 @@ def sample_inputs_linalg_vector_norm(op_info, device, dtype, requires_grad, **kw
 # specify scalars of floating, integral & complex types as values for "alpha".
 # Keyword argument `rhs_exclude_zero` is used to exclude zero values from rhs tensor argument
 # This is necessary for operations like `true_divide`, where divide by zero throws an exception.
-# Keyword argument `rounding_mode` is specifically used for the same kwarg supported by `torch.div`.
-def sample_inputs_binary_pwise(op_info, device, dtype, requires_grad, **kwargs):
+def sample_inputs_binary_pwise(op_info, device, dtype, requires_grad, extra_kwargs=None, **kwargs):
+    if extra_kwargs is None:
+        extra_kwargs = {}
+
     scalar = 3.14 + 3.14j if dtype.is_complex else (3.14 if dtype.is_floating_point else 3)
     scalar = 1 if dtype is torch.bool else scalar
     tests_list = [
@@ -747,23 +749,20 @@ def sample_inputs_binary_pwise(op_info, device, dtype, requires_grad, **kwargs):
     for first_shape, shape_or_scalar, broadcasts_input in test_cases:
         arg = shape_or_scalar
 
-        sample_kwargs = None  # type: ignore[assignment]
-        rounding_mode = kwargs.get('rounding_mode', None)
-        if rounding_mode is not None:
-            sample_kwargs = dict(rounding_mode=rounding_mode)
         if isinstance(shape_or_scalar, tuple):
             exclude_zero = kwargs.get('rhs_exclude_zero', False)
             arg = make_tensor(shape_or_scalar, device=device, dtype=dtype,
                               requires_grad=requires_grad, exclude_zero=exclude_zero)
         samples.append(SampleInput(make_tensor(first_shape, device=device, dtype=dtype,
                                                requires_grad=requires_grad),
-                                   args=(arg,), kwargs=sample_kwargs,
+                                   args=(arg,), kwargs=extra_kwargs,
                                    broadcasts_input=broadcasts_input))
     # Adds an extra sample using "alpha" if it's passed in kwargs
     if 'alpha' in kwargs:
         a = make_tensor((S, S, S), device=device, dtype=dtype, requires_grad=requires_grad)
         b = make_tensor((S, S, S), device=device, dtype=dtype, requires_grad=requires_grad)
-        sample = SampleInput(a, args=(b,), kwargs={'alpha': kwargs['alpha']})
+        extra_kwargs['alpha'] = kwargs['alpha']
+        sample = SampleInput(a, args=(b,), kwargs=extra_kwargs)
         samples.append(sample)
     return tuple(samples)
 
@@ -4567,12 +4566,12 @@ op_db: List[OpInfo] = [
     OpInfo('div',
            variant_test_name='true_rounding',
            dtypes=all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16),
-           sample_inputs_func=partial(sample_inputs_binary_pwise, rounding_mode=None, rhs_exclude_zero=True),
+           sample_inputs_func=partial(sample_inputs_binary_pwise, extra_kwargs={"rounding_mode":None}, rhs_exclude_zero=True),
            assert_autodiffed=True),
     OpInfo('div',
            variant_test_name='trunc_rounding',
            dtypes=all_types_and(torch.half, torch.bfloat16),
-           sample_inputs_func=partial(sample_inputs_binary_pwise, rounding_mode='trunc', rhs_exclude_zero=True),
+           sample_inputs_func=partial(sample_inputs_binary_pwise, extra_kwargs={"rounding_mode":'trunc'}, rhs_exclude_zero=True),
            skips=(
                # Reference: https://github.com/pytorch/pytorch/issues/59174
                SkipInfo('TestCommon', 'test_variant_consistency_jit'),
@@ -4581,7 +4580,7 @@ op_db: List[OpInfo] = [
     OpInfo('div',
            variant_test_name='floor_rounding',
            dtypes=all_types_and(torch.half, torch.bfloat16),
-           sample_inputs_func=partial(sample_inputs_binary_pwise, rounding_mode='floor', rhs_exclude_zero=True),
+           sample_inputs_func=partial(sample_inputs_binary_pwise, extra_kwargs={"rounding_mode":'floor'}, rhs_exclude_zero=True),
            skips=(
                # Reference: https://github.com/pytorch/pytorch/issues/59174
                SkipInfo('TestCommon', 'test_variant_consistency_jit'),
```

kshitij12345 replied:
Unfortunately, dict can't be default as it is mutable argument.

https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

Is it okay to rename `extra_args` as `sample_kwargs`?

mruberry replied:
Good catch, thank you

Sure, whatever name you like

mruberry reviewed (commented):
This looks pretty good @kshitij12345; I made a few comments inline for your review. Looking forward to porting one of the largest remaining set of method_tests()!

kshitij12345 committed:
address review

kshitij12345 committed:
Merge branch 'develop/opinfo/div' into develop/opinfo/div

kshitij12345 commented:
@mruberry Addressed the reviews. PTAL :)

Changes
* Add extra_kwargs
* Removed one redundant entry
* Updated the `summary` to have branch only for list and tuple

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -739,7 +732,10 @@ def sample_inputs_linalg_vector_norm(op_info, device, dtype, requires_grad, **kw
 # Keyword argument `rhs_exclude_zero` is used to exclude zero values from rhs tensor argument
 # This is necessary for operations like `true_divide`, where divide by zero throws an exception.
-def sample_inputs_binary_pwise(op_info, device, dtype, requires_grad, **kwargs):
+def sample_inputs_binary_pwise(op_info, device, dtype, requires_grad, extra_kwargs=None, **kwargs):
```
`extra_kwargs` is a better name

Style nit (only to be fixed if significant changes are required): make `extra_kwargs` keyword-only?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1263,20 +1261,6 @@ def sample_inputs_comparison_ops(self, device, dtype, requires_grad, **kwargs):
     sample_inputs = [*sample_inputs, *more_inputs]
     return tuple(sample_inputs)
 
-def sample_inputs_div(self, device, dtype, requires_grad, rounding_mode=None, **kwargs):
```
Nice simplification

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -4701,26 +4685,27 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
-           sample_inputs_func=sample_inputs_div,
-           skips=(SkipInfo('TestOpInfo', 'test_duplicate_method_tests'),),
-           assert_autodiffed=True),
-    OpInfo('div',
```
Nice job noticing this was redundant

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6864,24 +6849,6 @@ def ident(x):
 def method_tests():
     set_rng_seed(SEED)
     return [
-        ('div', (S, S, S), (torch.rand(S, S, S) + 0.1,), '', (True,)),
```
Wahoo!

mruberry reviewed (approved):
Awesome work, @kshitij12345!
